File: WitchcraftApp/src/test/java/com/tarbadev/witchcraft/weeks/domain/usecase/WeekFromYearAndWeekNumberUseCaseTest.java
package com.tarbadev.witchcraft.weeks.domain.usecase;

import com.tarbadev.witchcraft.weeks.domain.usecase.WeekFromYearAndWeekNumberUseCase;
import com.tarbadev.witchcraft.weeks.domain.entity.Day;
import com.tarbadev.witchcraft.weeks.domain.entity.DayName;
import com.tarbadev.witchcraft.recipes.domain.entity.Recipe;
import com.tarbadev.witchcraft.weeks.domain.entity.Week;
import com.tarbadev.witchcraft.weeks.domain.repository.WeekRepository;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.Mock;
import org.mockito.junit.MockitoJUnitRunner;

import java.util.Arrays;

import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.BDDMockito.given;

@RunWith(MockitoJUnitRunner.class)
public class WeekFromYearAndWeekNumberUseCaseTest {
  @Mock private WeekRepository weekRepository;

  private WeekFromYearAndWeekNumberUseCase subject;

  @Before
  public void setUp() {
    subject = new WeekFromYearAndWeekNumberUseCase(weekRepository);
  }

  @Test
  public void execute() {
    Week week = Week.builder()
        .weekNumber(12)
        .year(2018)
        .days(Arrays.asList(
            Day.builder().name(DayName.MONDAY).lunch(Recipe.builder().build()).diner(Recipe.builder().build()).build(),
            Day.builder().name(DayName.TUESDAY).lunch(Recipe.builder().build()).diner(Recipe.builder().build()).build(),
            Day.builder().name(DayName.WEDNESDAY).lunch(Recipe.builder().build()).diner(Recipe.builder().build()).build(),
            Day.builder().name(DayName.THURSDAY).lunch(Recipe.builder().build()).diner(Recipe.builder().build()).build(),
            Day.builder().name(DayName.FRIDAY).lunch(Recipe.builder().build()).diner(Recipe.builder().build()).build(),
            Day.builder().name(DayName.SATURDAY).lunch(Recipe.builder().build()).diner(Recipe.builder().build()).build(),
            Day.builder().name(DayName.SUNDAY).lunch(Recipe.builder().build()).diner(Recipe.builder().build()).build()
        ))
        .build();

    given(weekRepository.findByYearAndWeekNumber(week.getYear(), week.getWeekNumber())).willReturn(week);

    assertThat(subject.execute(week.getYear(), week.getWeekNumber())).isEqualTo(week);
  }

  @Test
  public void execute_databaseNullReturnsNewWeek() {
    int year = 2018;
    int weekNumber = 12;
    Week week = Week.builder()
        .year(year)
        .weekNumber(weekNumber)
        .days(Arrays.asList(
            Day.builder()
                .name(DayName.MONDAY)
                .lunch(Recipe.builder().build())
                .diner(Recipe.builder().build())
                .build(),
            Day.builder()
                .name(DayName.TUESDAY)
                .lunch(Recipe.builder().build())
                .diner(Recipe.builder().build())
                .build(),
            Day.builder()
                .name(DayName.WEDNESDAY)
                .lunch(Recipe.builder().build())
                .diner(Recipe.builder().build())
                .build(),
            Day.builder()
                .name(DayName.THURSDAY)
                .lunch(Recipe.builder().build())
                .diner(Recipe.builder().build())
                .build(),
            Day.builder()
                .name(DayName.FRIDAY)
                .lunch(Recipe.builder().build())
                .diner(Recipe.builder().build())
                .build(),
            Day.builder()
                .name(DayName.SATURDAY)
                .lunch(Recipe.builder().build())
                .diner(Recipe.builder().build())
                .build(),
            Day.builder()
                .name(DayName.SUNDAY)
                .lunch(Recipe.builder().build())
                .diner(Recipe.builder().build())
                .build()
        ))
        .build();

    given(weekRepository.findByYearAndWeekNumber(year, weekNumber)).willReturn(null);

    assertThat(subject.execute(year, weekNumber)).isEqualTo(week);
  }
}

File: WitchcraftApp/src/test/kotlin/com/tarbadev/witchcraft/weeks/domain/usecase/WeekFromYearAndWeekNumberUseCaseTest.kt
package com.tarbadev.witchcraft.weeks.domain.usecase

import com.nhaarman.mockitokotlin2.mock
import com.nhaarman.mockitokotlin2.whenever
import com.tarbadev.witchcraft.recipes.domain.entity.Recipe
import com.tarbadev.witchcraft.weeks.domain.entity.*
import com.tarbadev.witchcraft.weeks.domain.repository.WeekRepository
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test

class WeekFromYearAndWeekNumberUseCaseTest {
    private val weekRepository: WeekRepository = mock()
    private lateinit var weekFromYearAndWeekNumberUseCase: WeekFromYearAndWeekNumberUseCase

    @BeforeEach
    fun setUp() {
        weekFromYearAndWeekNumberUseCase = WeekFromYearAndWeekNumberUseCase(weekRepository)
    }

    @Test
    fun execute() {
        val week = Week(
            weekNumber = 12,
            year = 2018,
            days = listOf(
                Day(name = DayName.MONDAY, meals = listOf(Meal(mealType = MealType.LUNCH, recipe= Recipe()))),
                Day(name = DayName.TUESDAY, meals = listOf(Meal(mealType = MealType.DINER, recipe= Recipe()))),
                Day(name = DayName.WEDNESDAY, meals = listOf(Meal(mealType = MealType.LUNCH, recipe= Recipe()))),
                Day(name = DayName.THURSDAY, meals = listOf(Meal(mealType = MealType.LUNCH, recipe= Recipe()))),
                Day(name = DayName.FRIDAY, meals = listOf(Meal(mealType = MealType.DINER, recipe= Recipe()))),
                Day(name = DayName.SATURDAY, meals = listOf(Meal(mealType = MealType.LUNCH, recipe= Recipe()))),
                Day(name = DayName.SUNDAY, meals = listOf(Meal(mealType = MealType.DINER, recipe= Recipe())))
            ))

        whenever(weekRepository.findByYearAndWeekNumber(week.year, week.weekNumber)).thenReturn(week)

        assertEquals(week, weekFromYearAndWeekNumberUseCase.execute(week.year, week.weekNumber))
    }

    @Test
    fun execute_databaseNullReturnsNewWeek() {
        val year = 2018
        val weekNumber = 12
        val week = Week(
            year = year,
            weekNumber = weekNumber,
            days = listOf(
                Day(name = DayName.MONDAY),
                Day(name = DayName.TUESDAY),
                Day(name = DayName.WEDNESDAY),
                Day(name = DayName.THURSDAY),
                Day(name = DayName.FRIDAY),
                Day(name = DayName.SATURDAY),
                Day(name = DayName.SUNDAY)
            ))

        whenever(weekRepository.findByYearAndWeekNumber(year, weekNumber)).thenReturn(null)

        assertEquals(week, weekFromYearAndWeekNumberUseCase.execute(year, weekNumber))
    }
}